### PR TITLE
Salloum solution 2

### PR DESCRIPTION
**Pull request by MohammadRAlSalloum**
---
name: solution review 
about: A template PR for code review with a checklist
---

## Behavior

*This function takes a string as input then it outputs a list of all integers found in the input string.*

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [ ] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [ ] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations

### Do's

- [x] Variable names help to understand the strategy
- [ ] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [ ] Comments explain the _strategy_, **not** the _implementation_
- [ ] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/extract_integers.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
+"""
+a module that extract all the integers found in a given string.
+
+@author: Mohammad R Al Salloum.
+@date: 01/05/2025
+
+"""
+
+import re
```
Comment (AhmedKhalifa7) [inline solutions/extract_integers.py:9]: Just a question, what is re?

File: solutions/tests/test_extract_integers.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    code is implemented correctly and matches the expectations of its
+    behavior.
+    """
+
+    def test_positive_integer_in_string(self):
+        """
+        Checks if the function works with strings containing positive integers.
+        """
+        self.assertEqual(extract_integers("I have 3 apples"), [3])
+        self.assertEqual(extract_integers("I have 3 apples and 5 bananas"), [3, 5])
```
Comment (AhmedKhalifa7) [inline solutions/tests/test_extract_integers.py:38]: You have test with two assertions while Evan asked for one per test. You can either delete them or add more tests that you already have.

**Review by AhmedKhalifa7 — CHANGES_REQUESTED** [review]
Everything looks clean and amazing!, other than the test problem (and it's not a problem by the way you can really benefit from it) everything is great and well coded and documented!

**Review by kerthnorth — COMMENTED** [review]
The code seems to function well and handles multiple instances.

Great Job on documenting both the code and test cases. 

The tests consider multiple varying scenarios. This is helpful in helping us to better understand the behavior. Great Job!

File: solutions/extract_integers.py (hunk 1 above)
Comment (MohammadRAlSalloum) [reply solutions/extract_integers.py:9]: It stands for regular expression module; it searches, matches, and manipulates strings using patterns.

File: solutions/tests/test_extract_integers.py (hunk 2 above)
Comment (MohammadRAlSalloum) [reply solutions/tests/test_extract_integers.py:37]: Resolved

**Review by AhmedKhalifa7 — DISMISSED** [review]
Everything looks great, as best as it could be!
But I have a question, I see you push both of your solutions and tests in this PR, is it fine? If not you can delete them, if it is fine then ..... :shipit:  !!!

File: solutions/__init__.py
Comment (RaedAlyan) [inline solutions/__init__.py:1]: please don't push this file because I've pushed it before

File: solutions/guessing_game.py
Comment (RaedAlyan) [inline solutions/guessing_game.py:1]: why did you push this file again? It has been withiin another PR, right?

File: solutions/tests/test_guessing_game.py
Comment (RaedAlyan) [inline solutions/tests/test_guessing_game.py:1]: you've pushed this file before

File: solutions/tests/__init__.py
Comment (RaedAlyan) [inline solutions/tests/__init__.py:1]: please don't push this file, because I've pushed it before

File: solutions/__init__.py
Comment (MohammadRAlSalloum) [reply solutions/__init__.py:1]: If I remove it CI checks will not be completed; I removed it anyway

File: solutions/tests/test_guessing_game.py
Comment (MohammadRAlSalloum) [reply solutions/tests/test_guessing_game.py:1]: Reolved

File: solutions/tests/__init__.py
Comment (MohammadRAlSalloum) [reply solutions/tests/__init__.py:1]: CI checks won't be completed without it; I have removed it anyways
